hbase-server/src/main/java/org/apache/hadoop/hbase/master/DrainingServerTracker.java

```diff
@@ -77,36 +82,66 @@ public void serverAdded(ServerName sn) {
         }
       }
     });
+
     List<String> servers =
       ZKUtil.listChildrenAndWatchThem(watcher, watcher.getZNodePaths().drainingZNode);
-    add(servers);
+
+    if (servers != null) {
+      add(servers);
+    }
   }
 
   private void add(final List<String> servers) throws IOException {
     synchronized (this.drainingServers) {
-      this.drainingServers.clear();
+      // Clear all servers from the draining list that shouldn't stay drained. Information about
+      // whether a server should stay drained or not is added to the ZNode by the HMaster
+      this.drainingServers.removeIf(sn -> !shouldServerStayDrained(sn));
+
       for (String n : servers) {
         final ServerName sn = ServerName.valueOf(ZKUtil.getNodeName(n));
         this.drainingServers.add(sn);
         this.serverManager.addServerToDrainList(sn);
-        LOG.info("Draining RS node created, adding to list [" + sn + "]");
-
+        LOG.info("Draining RS node created, adding to list [{}]", sn);
       }
     }
   }
 
   private void remove(final ServerName sn) {
     synchronized (this.drainingServers) {
+      if (shouldServerStayDrained(sn)) {
+        LOG.info(
+          "Refusing to remove drained RS {} from the list, it's marked as permanently drained", sn);
+        return;
+      }
       this.drainingServers.remove(sn);
       this.serverManager.removeServerFromDrainList(sn);
+      LOG.info("Successfully removed drained RS {} from the list", sn);
     }
   }
 
+  private boolean shouldServerStayDrained(final ServerName sn) {
+    boolean shouldBePermanentlyDecommissioned = false;
+    String parentZnode = this.watcher.getZNodePaths().drainingZNode;
+    String node = ZNodePaths.joinZNode(parentZnode, sn.getServerName());
+
+    try {
+      byte[] rawData = ZKUtil.getData(this.watcher, node);
+      // Check if the data is present for backwards compatibility, some nodes may not have it
+      if (rawData != null && rawData.length > 0) {
+        DrainedZNodeServerData znodeData = DrainedZNodeServerData.parseFrom(rawData);
+        shouldBePermanentlyDecommissioned = znodeData.getMatchHostNameOnly();
+      }
+    } catch (InterruptedException | KeeperException | InvalidProtocolBufferException e) {
+      // pass
```

Comment:
For InterruptedException, you should propagate it as far as you can. InvalidProtocolBufferException should be logged at warn. KeeperException, yeah probably propagate that too -- there's nothing gained by trying to hide the fact from itself that the DrainingServerTracker is using zookeeper